core/Symbols.cc

```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
     return result;
 }
 
+// Record a required ancestor for this class of module
+void Symbol::recordRequiredAncestor(MutableContext ctx, SymbolRef ancestor, Loc loc) {
+    ENFORCE(this->isClassOrModule(), "Symbol is not a class or module: {}", this->show(ctx));
+
+    // We store the required ancestors into a fake property called `<required-ancestors>`
+    auto ancestors = this->findMember(ctx, Names::requiredAncestors());
+    if (!ancestors.exists()) {
+        ancestors = ctx.state.enterMethodSymbol(loc, this->ref(ctx), Names::requiredAncestors());
+        ancestors.data(ctx)->locs_.clear(); // Remove the original location
+    }
+
+    if (ancestors.data(ctx)->resultType == nullptr) {
+        vector<TypePtr> targs;
+        ancestors.data(ctx)->resultType = TupleType::build(ctx, move(targs));
+    } else {
+        // Do not require the same ancestor twice
+        for (auto elem : (cast_type<TupleType>(ancestors.data(ctx)->resultType))->elems) {
+            ENFORCE(isa_type<ClassType>(elem), "Something in requiredAncestors that's not a ClassType");
+            if (cast_type_nonnull<ClassType>(elem).symbol == ancestor) {
+                return;
+            }
+        }
+    }
+
+    auto type = core::make_type<ClassType>(ancestor);
+    (cast_type<TupleType>(ancestors.data(ctx)->resultType))->elems.emplace_back(type);
+    ancestors.data(ctx)->locs_.emplace_back(loc);
+}
+
+// Locally required ancestors by this class or module
+vector<Symbol::RequiredAncestor> Symbol::requiredAncestors(const GlobalState &gs) const {
+    ENFORCE(this->isClassOrModule(), "Symbol is not a class or module: {}", this->show(gs));
+
+    vector<RequiredAncestor> res;
+
+    auto ancestors = this->findMember(gs, Names::requiredAncestors());
+    if (!ancestors.exists()) {
+        // No ancestor was recorded for this class or module
+        return res;
+    }
+
+    auto data = ancestors.data(gs);
+    auto types = cast_type<TupleType>(data->resultType);
+    auto index = 0;
+    for (auto elem : types->elems) {
+        ENFORCE(isa_type<ClassType>(elem), "Something in requiredAncestors that's not a ClassType");
+        ENFORCE(index < data->locs().size(), "Missing loc in requiredAncestors");
+
+        RequiredAncestor req;
+        req.symbol = cast_type_nonnull<ClassType>(elem).symbol;
+        req.loc = data->locs().at(index);
```

Comment:
The loc the ancestor was required at (ie the loc of the `requires_ancestor X`) is at the same index than the `ANCST` type in the returned tuple.